storage/remote/queue_manager.go

```diff
@@ -1074,8 +1162,8 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface
 		case <-ctx.Done():
 			// In this case we drop all samples in the buffer and the queue.
 			// Remove them from pending and mark them as failed.
-			droppedSamples := nPendingSamples + int(s.enqueuedSamples.Load())
-			droppedExemplars := nPendingExemplars + int(s.enqueuedExemplars.Load())
+			droppedSamples := int(s.enqueuedSamples.Load())
+			droppedExemplars := int(s.enqueuedExemplars.Load())
```

Comment:
Should this take account of what is in `queue.Batch()` ?

Reply:
I don't think so, and I think this code is broken right now actually. Let me double check with you though:
1. We add to `enqueued{Samples,Exemplars}` during an append
2. We *should* subtract after sending a batch (I don't actually see this happening here or in main...)
3. If an error happens (hard shutdown) we will increment `failed{Samples,Exemplars}Total` in `s.sendSamples` for the current batch, then decrement `pending{Samples,Exemplars}` as if it was a sent batch.

Since will not subtract until after we send I think the current batch is accounted for by `s.sendSamples` failing and then subtracting/appending properly and everything else is still present in in the values being loaded.

Once we decide on which approach I will fix the bug about not ever decrementing right now...

Reply:
I think it should take the pending batch in, that's essentially what `nPendingType` were used for. This is a hard shutdown case and we're losing everything in the channel and everything already buffered in the shards batch.

Reply:
I think I phrased my response poorly. The batch being processed is taken into account, but via the code here: https://github.com/prometheus/prometheus/pull/9830/files#diff-1e66d5e658a9d99e1bd9b7be7d626d6b3bf9bbf480e210d2946c562cd746767cR1252-R1257. When a hard shutdown occurs that code is still executed before we enter the `<-ctx.Done()` case so the subtractions will already be done.